gallery_dl/extractor/discord.py

```diff
@@ -0,0 +1,370 @@
+    directory_fmt = (
+        "{category}", "{server_id}_{server}", "{channel_id}_{channel}"
+    )
+    pattern = BASE_PATTERN + r"/channels/(\d+)/?$"
+    example = (
+        "https://discord.com/channels/302094807046684672"
+    )
+
+    def items(self):
+        server_id = self.groups[0]
+
+        self.build_server_and_channels(server_id)
+
+        for channel in self.server_channels_metadata.copy().values():
+            if channel["channel_type"] in (0, 5, 15, 16):
+                yield from self.extract_channel(
+                    channel["channel_id"], safe=True
+                )
+
+
+class DiscordDirectMessagesExtractor(DiscordExtractor):
+    subcategory = "direct-messages"
+    directory_fmt = (
+        "{category}", "{subcategory}", "{channel_id}_{recipients:J,}"
+    )
+    pattern = BASE_PATTERN + r"/channels/@me/(\d+)/?$"
+    example = (
+        "https://discord.com/channels/@me/302094807046684672"
+    )
+
+    def items(self):
+        channel_id = self.groups[0]
+
+        yield from self.extract_channel(channel_id)
+
+
+class DiscordAPI():
+    """Interface for the Discord API v10
+
+    https://discord.com/developers/docs/reference
+    """
+
+    def __init__(self, extractor):
+        self.extractor = extractor
+        self.token = extractor.token
+        self.root = extractor.root + "/api/v10"
+
+    def get_server(self, server_id):
+        """Get server information"""
+        return self._call("/guilds/" + server_id)
+
+    def get_server_channels(self, server_id):
+        """Get server channels"""
+        return self._call("/guilds/" + server_id + "/channels")
+
+    def get_channel(self, channel_id):
+        """Get channel information"""
+        return self._call("/channels/" + channel_id)
+
+    def get_channel_threads(self, channel_id):
+        """Get channel threads"""
+        THREADS_BATCH = 25
+
+        def _method(offset):
+            return self._call(
+                "/channels/" + channel_id + "/threads/search?"
+                "sort_by=last_message_time&sort_order=desc"
+                "&limit=" + str(THREADS_BATCH) + "&offset=" + str(offset)
+            )["threads"]
+
+        return self._pagination(_method, THREADS_BATCH)
+
+    def get_channel_messages(self, channel_id):
+        """Get channel messages"""
+        MESSAGES_BATCH = 100
+
+        before = None
+
+        def _method(_):
+            nonlocal before
+            messages = self._call(
+                "/channels/" + channel_id +
+                "/messages?limit=" + str(MESSAGES_BATCH) +
+                (("&before=" + before) if before else "")
+            )
+            before = messages[-1]["id"]
+            return messages
+
+        return self._pagination(_method, MESSAGES_BATCH)
+
+    def _call(self, endpoint):
+        url = self.root + endpoint
+        try:
+            response = self.extractor.request(url, headers={
+                "Authorization": self.token,
+            })
```

Comment:
`headers` should be pre-computed in `__init__`
``` python
self.headers = {"Authorization": self.token}
```
```suggestion
            response = self.extractor.request(url, headers=self.headers)
```